### PR TITLE
Ensure error resolving an entity don't impact resolvable ones (for `main` branch)

### PR DESCRIPTION
**Pull request by pcmanus**
This PR has the same fix than in https://github.com/apollographql/federation/pull/1305, but for `main`.

_However_, testing this also uncovered another issue with `@requires` on `main` (fed 2 code), and this PR also has this additional fix. So first commit is that "other" fix (but it includes all the tests), and second commit is the fix of https://github.com/apollographql/federation/pull/1305.

### CODE REVIEW

**Comment by pcmanus** [comment]
@clenfest: the second commit here is going to be review as part of #1305/#376 and feel free to skip it (though certainly happy with more eyes on it), but the 1st commit here is specific to the fed2, and I think you might be the most qualified to review. Do you mind?

**Comment by clenfest** [comment]
Maybe it's because it's a little late in the day for me, but I think I would probably greatly benefit from a walk through of this code. Would you maybe have time tomorrow? Either way I'll dig in deeper tomorrow.

File: gateway-js/src/executeQueryPlan.ts [hunk 1]
```diff
@@ -489,7 +489,17 @@ function executeSelectionSet(
         const selections = (selection as QueryPlanFieldNode).selections;
 
         if (typeof source[responseName] === 'undefined') {
```
Comment (clenfest) [inline gateway-js/src/executeQueryPlan.ts:491]: should we simplify to 
```typescript
if (source[responseName] === undefined)
```

File: query-graphs-js/src/querygraph.ts [hunk 2]
```diff
@@ -353,6 +360,16 @@ export class QueryGraph {
     const indexes = this.typesToVertices.get(typeName);
     return indexes == undefined ? [] : indexes.map(i => this.vertices[i]);
   }
+
+  externalTester(source: string) {
```
Comment (clenfest) [inline query-graphs-js/src/querygraph.ts:364]: nit: return type

File: query-planner-js/src/buildPlan.ts [hunk 3]
```diff
@@ -1637,7 +1642,9 @@ function inputsForRequire(graph: QueryGraph, entityType: ObjectType, edge: Edge,
   fullSelectionSet.add(new FieldSelection(new Field(entityType.typenameField()!)));
   fullSelectionSet.mergeIn(edge.conditions!);
   if (includeKeyInputs) {
-    fullSelectionSet.mergeIn(additionalKeyEdgeForRequireEdge(graph, edge).conditions!);
+    const keyCondition = getLocallySatisfiableKey(graph, edge.head);
+    assert(keyCondition, () => `No key found for require on ${edge}`);
```
Comment (clenfest) [inline query-planner-js/src/buildPlan.ts:1646]: Maybe a slightly better error message? Suggestion: "Use of @require mandates a @key and no key was found on ${edge}"

**Review by clenfest — COMMENTED** [review]
Couple of super minor comments

File: gateway-js/src/executeQueryPlan.ts (hunk 1 above)
Comment (clenfest) [reply gateway-js/src/executeQueryPlan.ts:491]: Oops, did you not touch this line? Never mind.

**Review by clenfest — APPROVED** [review]

**Pull request merged**